LSPS0/README.md

```diff
@@ -0,0 +1,456 @@
+> embedded 0 bytes may cause problems in implementations
+> that pass the JSON string to C code.
+> Conversely, we do not require the payload to be terminated
+> by a 0 byte / NUL character as it is unnecessary in many
+> modern non-C-based languages; C code can copy the buffer
+> and append the NUL character if necessary, as the payload
+> size is known.
+
+The JSON object embedded in the message payload is defined
+by the [JSON-RPC 2.0][] protocol.
+
+[JSON-RPC 2.0]: https://www.jsonrpc.org/specification
+
+The LSP acts as a JSON-RPC 2.0 server, while the client acts
+as a JSON-RPC 2.0 client.
+
+> **Rationale** JSON is a simple format that is easy to
+> extend and is extensively supported.
+> The Lightning Network peer-to-peer transport protocol in
+> BOLT8 is not inherently a request-response protocol like
+> HTTP(S) or gRPC, and JSON-RPC 2.0 describes a
+> JSON-based protocol that builds a request-response
+> protocol from a simple tunnel; message ID 37913 acts as
+> that tunnel.
+> JSON-RPC 2.0 is simple to implement.
+> Although BOLT8 messages are limited to payloads of 65533
+> bytes, it is expected that both requests and responses
+> would be far below that limit, and thus the ability to
+> "cut" a large object across multiple messages, the
+> use of a compression algorithm, and the use of a binary
+> format instead of JSON, are all considered unnecessary.
+
+The client:
+
+* MUST send single complete JSON-RPC 2.0 request objects,
+  UTF-8-encoded, as payloads for message ID 37913.
+* MUST NOT send JSON-RPC 2.0 notification objects
+  (i.e. every object it sends must have an `id` field).
+* MUST NOT use by-position parameter structures, and MUST
+  use by-name parameter structures.
+* MUST NOT batch multiple JSON-RPC 2.0 request objects in
+  an array as described in the "Batch" section of the
+  JSON-RPC 2.0 specification.
+
+> **Rationale** By disallowing by-position parameter
+> structures, other LSPS specifications need only to define
+> parameter names and not some canonical order of parameters
+> for by-position use.
+> Having to handle only by-name parameter structures also
+> simplifies the LSP code, as it does not have to check
+> whether the `params` value is an array or a dictionary,
+> and to separately map array elements to `params` keys.
+> Batched requests require batched responses according to
+> JSON-RPC 2.0, and it may be simpler for LSPs to handle
+> unrelated LSPS methods separately without requiring
+> re-batching of the responses; it gives a simple "one
+> message is one request / response" rule.
+
+The LSP:
+
+* MUST send single complete JSON-RPC 2.0 response objects,
+  UTF-8-encoded, as payloads for message ID 37913.
+* MUST NOT send JSON-RPC 2.0 notification objects
+  (i.e. every object it sends must have an `id` field).
+* MUST respond with an error -32602 if a request `params` field
+  is not recognized for the given `method`.
+* MUST NOT send message ID 37913 unless the peer first sent a
```

Comment:
A thing is that if the client is a mobile device, it might be sent to sleep by the mobile device OS (which also disconnects its connection, I believe).  Then to wake it up, we need a mobile-device-OS-specific mechanism to notify them.  In some mobile device OSs, apparently this requires the private key of the mobile app developer, i.e. the developer of the *client*.  So we need some way to contact the client *developer* so it can go sign the notification and then notify the client so it can wake up, so maybe we could ride on an LSP-initiated conversation on that notification anyway.  In the TG it was agreed to defer this to after LSPS0,1,3 are ratified (the initial thought being that it would be effectively the client registers a webhook via some LSPS0-based protocol, then the LSP pings that webhook, but details TBD).

On the other hand, maybe some clients are desktop-only anyway.

A similar mechanism would be desired by async payments.  Indeed JIT channels and async payments would both require some kind of mechanism to have LSP initiate a conversation, mostly "hey someone wants to pay you!".

I am fine with allowing the LSP to emit a JSON-RPC 2.0 notification object (which has no `id` and cannot be responded to), provided the client with the same node ID has ever sent a 37913 in the past, possibly on a different connection session.  The expectation is that the client will first arrange an LSPS1 inbound channel, an LSPS3 JIT channel, or at the minimum queried the LSP for what protocols it knows via LSPS0 `listprotocols`.

Reply:
Allowed JSON-RPC 2.0 notifications from the LSP, but not from the client.  Also added prefixing convention for LSPS-defined LSP-sourced notifications as `n.lsps999.`.

Reply:
resolving conversation, as there seem to be no objections to having JSON-RPC 2.0 notifications in the LSP->client direction, and no strong push to allow JSON-RPC 2.0 requests initiated by the LSP.